File: kotlin-node/karakum/src/jsMain/resources/plugins/convertTypealiasParameterBounds.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        sourceFileName.endsWith("stream.d.ts")
        && ts.isTypeParameterDeclaration(node)
        && node.parent
        && ts.isTypeAliasDeclaration(node.parent)
        && (
            node.parent.name.text === "PipelineCallback"
            || node.parent.name.text === "PipelineDestination"
            || node.parent.name.text === "PipelinePromise"
            || node.parent.name.text === "PipelineTransform"
        )
    ) {
        const name = render(node.name)

        const constraintType = node.constraint && render(node.constraint)
        const defaultType = node.default && render(node.default)

        const bound = `${karakum.ifPresent(constraintType, it => ` : ${it}`)}${karakum.ifPresent(defaultType, it => ` default is ${it}`)}`

        return `${name}${karakum.ifPresent(bound, it => ` /* ${it} */`)}`
    }

    if (
        sourceFileName.endsWith("util.d.ts")
        && ts.isTypeParameterDeclaration(node)
        && node.parent
        && ts.isTypeAliasDeclaration(node.parent)
        && node.parent.name.text === "CustomPromisify"
    ) {
        const name = render(node.name)

        const constraintType = node.constraint && render(node.constraint)
        const defaultType = node.default && render(node.default)

        const bound = `${karakum.ifPresent(constraintType, it => ` : ${it}`)}${karakum.ifPresent(defaultType, it => ` default is ${it}`)}`

        return `${name}${karakum.ifPresent(bound, it => ` /* ${it} */`)}`
    }


    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertTypealiasParameterBounds.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.ifPresent
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isTypeAliasDeclaration
import typescript.isTypeParameterDeclaration

val convertTypealiasParameterBounds = createPlugin { node, context, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

        nullable {
            ensure(sourceFileName.endsWith("buffer.buffer.d.ts"))
            ensure(isTypeParameterDeclaration(node))

            val typeAlias = ensureNotNull(node.getParentOrNull())
            ensure(isTypeAliasDeclaration(typeAlias))
            ensure(typeAlias.name.text == "ImplicitArrayBuffer")

            val name = render(node.name)

            val constraintType = node.constraint?.let { render(it) }
            val defaultType = node.default?.let { render(it) }

            val bound = "${ifPresent(constraintType) { " : $it" }}${ifPresent(defaultType) { " default is $it" }}"

            "${name}${ifPresent(bound) { " /* $it */" }}"
        } ?: nullable {
            ensure(sourceFileName.endsWith("stream.d.ts"))

            ensure(isTypeParameterDeclaration(node))

            val typeAlias = ensureNotNull(node.getParentOrNull())
            ensure(isTypeAliasDeclaration(typeAlias))
            ensure(
                typeAlias.name.text == "PipelineCallback" ||
                        typeAlias.name.text == "PipelineDestination" ||
                        typeAlias.name.text == "PipelinePromise" ||
                        typeAlias.name.text == "PipelineTransform"
            )

            val name = render(node.name)

            val constraintType = node.constraint?.let { render(it) }
            val defaultType = node.default?.let { render(it) }

            val bound = "${ifPresent(constraintType) { " : $it" }}${ifPresent(defaultType) { " default is $it" }}"

            "${name}${ifPresent(bound) { " /* $it */" }}"
        } ?: nullable {
            ensure(sourceFileName.endsWith("util.d.ts"))

            ensure(isTypeParameterDeclaration(node))

            val typeAlias = ensureNotNull(node.getParentOrNull())
            ensure(isTypeAliasDeclaration(typeAlias))
            ensure(typeAlias.name.text == "CustomPromisify")

            val name = render(node.name)

            val constraintType = node.constraint?.let { render(it) }
            val defaultType = node.default?.let { render(it) }

            val bound = "${ifPresent(constraintType) { " : $it" }}${ifPresent(defaultType) { " default is $it" }}"

            "${name}${ifPresent(bound) { " /* $it */" }}"
        }
    }
}
